src/tseda/datastore.py

```diff
@@ -36,15 +38,186 @@ def preprocess(tsm):
+    formatters = {
+        "color": {"type": "color"},
+        "predefined": {"type": "tickCross"},
+    }
+
+    create_sample_set_textinput = param.String(
+        doc="Enter name of new sample set. Press Enter (⏎) to create.",
+        default=None,
+        label="Create new sample set",
+    )
+
+    sample_set_warning = pn.pane.Alert(
+        "This sample set name already exists, pick a unique name.",
+        alert_type="warning",
+        visible=False,
+    )
+
+    page_size = param.Selector(objects=[10, 20, 50, 100], default=20)
+
+    table = param.DataFrame()
+
+    def __init__(self, **params):
+        super().__init__(**params)
+        self.table.set_index(["sample_set_id"], inplace=True)
+        self.data = self.param.table.rx()
+
+    @property
+    def tooltip(self):
+        return pn.widgets.TooltipIcon(
+            value=(
+                "The name and color of each sample set are editable. In the "
+                "color column, select a color from the dropdown list. In the "
+                "individuals table, you can assign individuals to sample sets."
+            ),
+        )
+
+    @pn.depends("page_size", "create_sample_set_textinput")  # , "columns")
+    def __panel__(self):
+        if self.create_sample_set_textinput is not None:
+            previous_names = [
+                self.table.name[i] for i in range(len(self.table))
+            ]
+            if self.create_sample_set_textinput in previous_names:
+                self.sample_set_warning.visible = True
+            else:
+                previous_colors = [
+                    self.table.color[i] for i in range(len(self.table))
+                ]
+                unused_colors = [
+                    color
+                    for color in config.COLORS
+                    if color not in previous_colors
+                ]
+                if len(unused_colors) != 0:
+                    colors = unused_colors
+                else:
+                    colors = config.COLORS
+                self.sample_set_warning.visible = False
+                i = max(self.param.table.rx.value.index) + 1
+                self.param.table.rx.value.loc[i] = [
+                    self.create_sample_set_textinput,
+                    colors[random.randint(0, len(colors) - 1)],
+                    False,
+                ]
+                self.create_sample_set_textinput = None
+        table = pn.widgets.Tabulator(
+            self.data,
+            layout="fit_data_table",
+            selectable=True,
+            page_size=self.page_size,
+            pagination="remote",
+            margin=10,
+            formatters=self.formatters,
+            editors=self.editors,
+        )
+        return pn.Column(
+            pn.pane.Markdown("### Sample Set Table"), self.tooltip, table
+        )
+
+    def get_ids(self):
+        if isinstance(self.table, pd.DataFrame):
+            return [i for i in range(len(self.table["name"].tolist()))]
```

Comment:
This solution is not ideal, but since the table does not actually contain the indices this was a quick fix

Reply:
The index values should be available through `df.index.values` or something like that I think.

Reply:
Ah thank you!